src/nodes.coffee

```diff
@@ -2247,8 +2250,16 @@ exports.Assign = class Assign extends Base
 
       restElements
 
-    # Cache the value for reuse with rest elements
-    [@value, valueRef] = @value.cache o
+    # Cache the value for reuse with rest elements.
+    # `Obj` should be always cached.
+    # Examples:
+    #   {a, r...} = {a:1, b:2, c:3}
+    #   {a, r...} = {a:1, obj...}
+    shouldCache = (value) ->
+      return yes if value.base instanceof Obj
```

Comment:
Should we change `Obj::shouldCache` to always be true? I guess there may be unintended consequences of that, but I can't think of a situation where we'd rather duplicate an object literal rather than reusing it.